### PR TITLE
Features from MT Marathon

### PR DESCRIPTION
**Pull request by jindrahelcl**
- FactoredEncoder (can have n input data series - this should become the encoder implementation, maintaining SentenceEncoder as a special case with 1 input)
- MultiDecoder (has n child decoders, alternates between them)
- SequenceLabeler (a type of decoder for same length output e.g. for tagging)
- VanillaSentenceEncoder (SentenceEncoder without unused features and some hacks, maybe this should be deleted after inspecting those hacks)
- lots of new ini files in examples
- source output is written during validation

### CODE REVIEW

**Comment by jindrahelcl** [comment]
Before merge, we need to repair a function in the checking module which has been commented out.

File: examples/factored.ini [hunk 1]
```diff
@@ -0,0 +1,131 @@
+; This is an example configuration for training machine translation.  It is an
```
Comment (jlibovicky) [inline examples/factored.ini]: We probably don't need the same introduction in every ini file.

Comment (jlibovicky) [inline examples/factored.ini]: Instead, there should be what is this experiment supposed to do.

File: examples/tagging.ini [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+; in the INI file and can be referenced using the <> notation.
+;
+
+[main]
+; The main block contains the mandatory fields for running an experiment.
+output=out-example-tagging
+encoders=[<encoder>]
+decoder=<decoder_tags>
+runner=<runner>
+evaluation=[<bleu1>, <bleu4>, <acc>]
```
Comment (jlibovicky) [inline examples/tagging.ini]: Are you really measuring BLEU score on sequence labelling?

File: neuralmonkey/decoders/multi_decoder.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+import numpy as np
+
+from neuralmonkey.vocabulary import START_TOKEN
+from neuralmonkey.decoding_function import attention_decoder
+from neuralmonkey.logging import log
+from neuralmonkey.dataset import Dataset
+
+
+class MultiDecoder(object):
+    """A class that manages parts of the computation graph that are
```
Comment (jlibovicky) [inline neuralmonkey/decoders/multi_decoder.py]: Here, I would expect a thorough explanation what this class does and why would anyone need that.

File: neuralmonkey/decoders/multi_decoder.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    # Some locals may be turned to attributes
+
+    def __init__(self, main_decoder, regularization_decoders, **kwargs):
+        """Create a new instance of the multi-decoder.
+
+        Arguments:
+            main_decoder: The decoder that corresponds to the output which
+                we want at runtime.
+
+            additional_decoders: A list of the decoders among which the multidecoder
```
Comment (jlibovicky) [inline neuralmonkey/decoders/multi_decoder.py]: Name in documentation does not match the name in code.

File: neuralmonkey/decoders/multi_decoder.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        self.main_decoder = main_decoder
+
+        self.regularization_decoders = regularization_decoders
+
+        self._training_decoders = [main_decoder] + regularization_decoders
+        self._decoder_costs = tf.concat(0, [tf.expand_dims(d.cost, 0)
+                                            for d in self._training_decoders])
+
+        self._scheduled_decoder = 0
+        self._input_selector = tf.placeholder(tf.float32,
```
Comment (jlibovicky) [inline neuralmonkey/decoders/multi_decoder.py]: What is this for? (I mean, I can figure it out from the rest of the code, but it should be written here.)

File: neuralmonkey/decoders/sequence_labeler.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import tensorflow as tf
+
+from neuralmonkey.logging import log, debug
+from neuralmonkey.decoders.decoder import Decoder
+
+class SequenceLabeler(Decoder):
+
```
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_labeler.py]: Docstring

File: neuralmonkey/encoders/vanilla_sentence_encoder.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+from neuralmonkey.logging import log
+from neuralmonkey.nn.bidirectional_rnn_layer import BidirectionalRNNLayer
+from neuralmonkey.checking import assert_type
+from neuralmonkey.decoding_function import Attention
+from neuralmonkey.vocabulary import Vocabulary
+
+# tests: mypy
+
+class VanillaSentenceEncoder(object):
+    """Like the SentenceEncoder, but ignores the start/end tokens.
```
Comment (jlibovicky) [inline neuralmonkey/encoders/vanilla_sentence_encoder.py]: Why is this a separate class? Shouldn't it be rather an option in the standard encoder?

File: neuralmonkey/vocabulary.py [hunk 8]
```diff
@@ -363,6 +364,46 @@ def sentences_to_tensor(
         return word_indices, weights
 
 
+    def sentences_to_bare_tensor(
```
Comment (jlibovicky) [inline neuralmonkey/vocabulary.py]: Again, it should be an argument in a method, not an almost identical copy of it.

File: neuralmonkey/encoders/vanilla_sentence_encoder.py (hunk 7 above)
Comment (jindrahelcl) [inline neuralmonkey/encoders/vanilla_sentence_encoder.py]: Já bych to možná celý zahodil ten vanilla encoder.. My jsme nevěděli, jestli to náhodou neni bug. Ale pak se to učilo nějak debilně.. Nevíš, jestli se do enkodéru ty start a end tokeny maj nebo nemaj dávat? (Default co tam máme od začátku je že tam jsou, ale já necejtim moc důvod)

**Comment by jlibovicky** [comment]
Hm... mylsle jsem si, že approvuju jenom ty poslední změny, ne celej PR :-)

File: examples/tagging.ini [hunk 9]
```diff
@@ -0,0 +1,129 @@
+; This is an example configuration for training machine translation.  It is an
```
Comment (jlibovicky) [inline examples/tagging.ini]: This should also have a nice header as the ` examples/factored.ini ` - it is not an example configuration for MT.

File: tests/data/multi/train.forms-cs.txt [hunk 10]
```diff
@@ -0,0 +1,100 @@
+Jednašedesátiletý Pierre Vinken se připojí ke správní radě jako nevýkonný ředitel dne 29 . listopadu .
```
Comment (jlibovicky) [inline tests/data/multi/train.forms-cs.txt]: Ideally, we should add a README file describing the data (source, licence, preprocessing).

File: tests/data/train.tc.de [hunk 11]
```diff
@@ -0,0 +1,1000 @@
+zwei junge weiße Männer sind im Freien in der Nähe vieler Büsche .
```
Comment (jlibovicky) [inline tests/data/train.tc.de]: Wasn't there exactly the same data in a different PR?

**Pull request merged**